h/storage.py

```diff
@@ -217,37 +218,54 @@ def update_annotation(request, id_, data, group_service):
     return annotation
 
 
-def expand_uri(session, uri):
+def expand_uri(session, uri, normalized=False):
     """
     Return all URIs which refer to the same underlying document as `uri`.
 
     This function determines whether we already have "document" records for the
     passed URI, and if so returns the set of all URIs which we currently
     believe refer to the same document.
 
-    :param session: the database session
-    :type session: sqlalchemy.orm.session.Session
-
-    :param uri: a URI associated with the document
-    :type uri: str
+    :param session: Database session
+    :param uri: URI associated with the document
+    :param normalized: Return normalized URIs instead of the raw value
 
     :returns: a list of equivalent URIs
-    :rtype: list
     """
-    doc = models.Document.find_by_uris(session, [uri]).one_or_none()
 
-    if doc is None:
-        return [uri]
+    canonical_url = normalize_uri(uri)
+
+    document_id = (
+        session.query(models.DocumentURI.document_id)
+        .filter(models.DocumentURI.uri_normalized == canonical_url)
+        .limit(1)
+        .subquery()
+    )
+
+    type_uris = list(
+        session.query(
+            # Using the specific fields we want prevents object creation
+            # which significantly speeds this method up (knocks ~40% off)
+            models.DocumentURI.type,
+            models.DocumentURI.uri,
+            models.DocumentURI.uri_normalized,
+        ).filter(models.DocumentURI.document_id == document_id)
+    )
+
+    if not type_uris:
+        return [canonical_url if normalized else uri]
 
     # We check if the match was a "canonical" link. If so, all annotations
     # created on that page are guaranteed to have that as their target.source
     # field, so we don't need to expand to other URIs and risk false positives.
-    docuris = doc.document_uris
-    for docuri in docuris:
-        if docuri.uri == uri and docuri.type == "rel-canonical":
-            return [uri]
+    for doc_type, plain_uri, _ in type_uris:
+        if doc_type == "rel-canonical" and plain_uri == uri:
```

Comment:
Is this a bug? Should it be:

```suggestion
    for doc_type, _, normalized_uri in type_uris:
        if doc_type == "rel-canonical" and normalized_uri == canonical_uri:
```

Reply:
After further testing I've found that this block actually causes a problem. The comment says:

```py
# We check if the match was a "canonical" link. If so, all annotations
# created on that page are guaranteed to have that as their target.source
# field, so we don't need to expand to other URIs and risk false positives.
```

In the context of annotation search, the client will pass a list of URIs (known as the "search URIs") to the `/api/search` API. Each of these will be passed to `expand_uri` separately, so if one of the URIs is a `rel="canonical"` URI then that only affects the expansion of that query URI.

In the context of the WebSocket however the matching is done the other way around - the client passes these same two search URIs in the WS configuration. On the backend the annotation's `target_uri` is expanded using `expand_uri` when an annotation notification is delivered and the resulting set is checked against all the search URIs. If the `target_uri` corresponds to the `<link rel="canonical">` URL, then the expanded set will only include the `target_uri` and hence any clients viewing the document at a different URL will not receive an update.

Locally I had a situation where one of my test PDFs had a `<link rel="canonical">` URI in its set of equivalent URIs. I think this may have stemmed from a bug a few weeks ago where a buggy client build briefly treated PDF.js as an ordinary HTML document and the client's dev server includes a `<link rel="canonical">` tag in its PDF viewer that points to the original PDF URL (as opposed to any proxied URL from which the PDF viewer might have actually fetched the document). This situation shouldn't generally happen for proxied PDFs in production, but there might be some entries in the `document_uri` table for some PDFs if a similar situation has ever happened.

Reply:
The problem in the previous comment happened as a result of a bug, but a real issue that could happen is when two clients are viewing the same document at different URLs that are equivalent according to the data in the `document_uri` table, but only one of those URLs has `<link rel="canonical" ...>`. In this case any annotations made/updated on the canonical URL would not be reflected to clients at the equivalent-but-not-canonical URL.

Assuming my analysis is correct here, then I think we can afford not to make any changes to the `rel="canonical"` handling in this PR as it carries some risk and isn't needed to fix the issue at hand.

Reply:
That suits me. I'm very unclear on what this is attempting to do.